packages/const-serialize/README.md

```diff
@@ -0,0 +1,57 @@
+A rust serialization library that works in const with complex(ish) types like enums, nested structs and arrays. Const rust does not have an allocator, so this library cannot work in a cross architecture environment with Vecs, slices or strings.
+
+```rust
+use const_serialize::{deserialize_const, serialize_const, serialize_eq, ConstVec, SerializeConst};
+#[derive(Clone, Copy, Debug, PartialEq, SerializeConst)]
+struct Struct {
+    a: u32,
+    b: u8,
+    c: u32,
+    d: Enum,
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, SerializeConst)]
+#[repr(C, u8)]
+enum Enum {
+    A { one: u32, two: u16 },
+    B { one: u8, two: u16 } = 15,
+}
+
+const {
+    let data = [Struct {
+        a: 0x11111111,
+        b: 0x22,
+        c: 0x33333333,
+        d: Enum::A {
+            one: 0x44444444,
+            two: 0x5555,
+        },
+    }; 3];
+    let mut buf = ConstVec::new();
+    buf = serialize_const(&data, buf);
+    let buf = buf.read();
+    let (buf, deserialized) = match deserialize_const!([Struct; 3], buf) {
+        Some(data) => data,
+        None => panic!("data mismatch"),
+    };
+    if !serialize_eq(&data, &deserialized) {
+        panic!("data mismatch");
+    }
+}
+```
+
+## How it works
+
+`const-serialize` relies heavily on well defined layouts for the types you want to serialize. The serialization format is the linear sequence of unaligned bytes stored in the order of the fields, items or variants of the type. Numbers are stored in little endian order.
```

Comment:
Highly recommend reading this before reviewing const-serialize. Const rust is very limited which makes some of the code look odd. The serialization framework ends up looking very similar to zero copy serialization frameworks like:
- postcard-forth (a reimplementation of postcard serialization by the original author): https://onevariable.com/blog/pods-from-scratch
- zerocopy: https://github.com/google/zerocopy
Unlike zero copy serialization, const serialize normalizes endianness across platforms